[FSDP][1/n] Support LoRA training for FSDP backend.

GuanxingLu opened the pull request:
Approaches Considered:

1. Merge/Unmerge: Merge LoRA to base -> send merged weights -> unmerge for training.
2. File-based (Current): Save adapter to disk -> rollout engine loads via load_lora_adapter (e.g., AReal).
3. Direct Tensor: Send tensors via memory (e.g., verl).

Current Implementation: Implemented Option 2 using UpdateWeightFromDisk.

Limitation: When rollout engine offloading is enabled, the base model should be re-transmitted in each update (though overhead is manageable).

Status:
✅ Train Qwen3-4B with LoRA on single A100 40G (Full param triggers OOM).
🚧 Reward curve pending.

TODO:
- [ ] Implement Options 1 & 3 for comparison.

## Update
> Current Status:
> 
> * Add '--offload-rollout-level' argument, level 1: offload kv cache/cuda graph only, level 2: offload weights + kv cache/cuda graph. When the rollout engine  does not release the base model weights, we only need to sync LoRA weights.
> * I modified `UpdateWeight` base class (this should be invasive) to support LoRA update alongside base model weight update.
> * Current LoRA weight update logic works for both colocated and disaggregated modes.
> * Fixed nits you mentioned before. But I can not remove _lora_loaded because unload_lora_adapter will throw error if LoRA is not loaded.
> 
> More TODOs: 
- [ ] Save/Load logic. 
- [ ] Efficient layer summon for LoRA weight update. 
- [ ] Option 3: Send LoRA tensors rather than files.

Discussion Feedback on the trade-offs of these approaches is welcome!

PopSoda2002 reviewed (commented):
Has some nit for you :)

GuanxingLu committed:
Add LoRA for FSDP backend. (#416)

Co-authored-by: PopSoda2002 <zhouhp.me@gmail.com>

GuanxingLu commented:
@PopSoda2002 Could you please check this? Thank you!

Current Status:
- Add '--offload-rollout-level' argument, level 1: offload kv cache/cuda graph only, level 2: offload weights + kv cache/cuda graph. When the rollout engine  does not release the base model weights, we only need to sync LoRA weights.
- I modified `UpdateWeight` base class (this should be invasive) to support LoRA update alongside base model weight update.
- Current LoRA weight update logic works for both colocated and disaggregated modes. 
- Fixed nits you mentioned before. But I can not remove _lora_loaded because unload_lora_adapter will throw error if LoRA is not loaded.

More TODOs:
[] Save/Load logic.
[] Efficient layer summon for LoRA weight update.
[] Option 3: Send LoRA tensors rather than files.

PopSoda2002 commented:
Can you have some test case in wandb compared with no lora? I can help with this part

GuanxingLu commented:
Sure, you can run the test. I am also running them in my local setting.

GuanxingLu commented:
Sorry, now `pre-commit` tests are passed.

GuanxingLu committed:
Add LoRA save/load logic.

PopSoda2002 commented on slime/backends/fsdp_utils/actor.py:
```diff
@@ -134,11 +141,10 @@ def init(self, args: Namespace, role: str, with_ref: bool = False) -> int:  # ty
+        if self.args.colocate:
+            self.weight_updater = UpdateWeightFromTensor(self.args, self.model)
+        else:
+            self.weight_updater = UpdateWeightFromDistributed(self.args, self.model)
```
Can I ask why change this?

GuanxingLu committed:
Add LoRA: fixed weight update misalignment

arminZolfaghari commented:
Hi @GuanxingLu, thanks for working on this. 

I'm planning to run LoRA RL on Qwen3-8B and 14B with the `examples/retool` workflow on 4×GH200 (full-param OOMs on this box). Two questions:

1. Has anyone tried this branch with `examples/retool`, or only the math GRPO scripts?
2. Of the remaining TODOs (save/load, efficient layer summon, tensor-based sync), which are planned for `[2/n]` vs later PRs?

Happy to test the branch on 4×GH200 and report back if that's useful. 
Thanks!
